### PR TITLE
migrate existing integration tests from core + get it running on darwin/windows

### PR DESCRIPTION
**Pull request by jochil**
- building existing fuzz targets without bazel
- migrate `fuzzer-runner` tests
- get everything running on macosx and windows

### CODE REVIEW

File: integration/utils/runner.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	"code-intelligence.com/cifuzz/internal/config"
+	"code-intelligence.com/cifuzz/pkg/report"
+	"code-intelligence.com/cifuzz/pkg/runner/libfuzzer"
+	"code-intelligence.com/cifuzz/util/fileutil"
+	"code-intelligence.com/cifuzz/util/stringutil"
+	"github.com/stretchr/testify/require"
+)
+
+// ChannelPassthrough pipes the reports from the runner package to a report channel
+// and also print them to stdout
```
Comment (fmeum) [inline integration/utils/runner.go:21]:
```suggestion
// and also prints them to stdout
```

File: testdata/Makefile [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+				 -g
+
+#LDFLAGS = $(CFLAGS)
+
+BUILD_DIR = build
+FUZZ_TARGET_DIR = fuzz-targets
+SHARED_DIR = shared
+
+all:
+	-mkdir $(BUILD_DIR)
```
Comment (fmeum) [inline testdata/Makefile:24]: The indentation is a mix of spaces and tabs. We should reformat before importing.

File: testdata/Makefile [hunk 3]
```diff
@@ -0,0 +1,48 @@
+CC = clang
+CXX = clang++ 
+
+CFLAGS = -fsanitize=fuzzer,address,undefined \
+				 -fno-sanitize=function,vptr \
+				 -fsanitize-recover=address \
+				 -fsanitize-address-use-after-scope \
+				 -g
+				 #-static
```
Comment (fmeum) [inline testdata/Makefile:9]: I would prefer to delete or uncomment before importing. Also, we should probably synchronize this flag set with that used by the make integration.

File: testdata/shared/api.cpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+// Implementation of "api".
+#include "api.h"
+
+#include <iostream>
+#include <vector>
+
+// Do some computations with 'str', return the result.
+// This function contains a bug. Can you spot it?
+size_t DoStuff(const std::string &str) {
+  std::vector<int> Vec({0, 1, 2});
```
Comment (fmeum) [inline testdata/shared/api.cpp:13]: This seems to assume C++11, so we probably have to add `-std=c++11` to the Makefile.

File: testdata/Makefile (hunk 3 above)
Comment (jochil) [reply testdata/Makefile:9]: >  make integration

Do you mean cmake? If yes... any proposal how to do it?

File: testdata/Makefile (hunk 2 above)
Comment (jochil) [reply testdata/Makefile:22]: should be fixed now

File: integration/utils/report.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+		if options.Details != "" {
+			require.Contains(t, finding.GetDetails(), options.Details)
+		}
+
+		if !options.AllowEmptyInputData {
+			require.NotEmpty(t, finding.InputData, "InputData is empty")
+		}
+
+		// TODO is this necessary?
```
[the remark this answers is not in the document]
Comment (jochil) [reply integration/utils/report.go:47]: It is a new one... maybe @adombeck can answer this?
The `MoreDetails` field seems to be empty in the new `report` package unlike how it was in the `types` package

File: integration/utils/runner.go (hunk 1 above)
Comment (jochil) [reply integration/utils/runner.go]: done

File: integration/utils/report.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (adombeck) [reply integration/utils/report.go:47]: The field is still being set in some cases (https://github.com/CodeIntelligenceTesting/cifuzz/blob/333149579075806ae29ed92c3e08ced4bdb0e60c/pkg/parser/libfuzzer/libfuzzer_parser.go#L312-L312, https://github.com/CodeIntelligenceTesting/cifuzz/blob/333149579075806ae29ed92c3e08ced4bdb0e60c/pkg/parser/libfuzzer/libfuzzer_parser.go#L399-L399), but not in all cases as in the core repo, so I think we can just skip this check here.

[the remark this answers is not in the document]
Comment (jochil) [reply integration/utils/report.go]: ok, I removed the check/comment

File: testdata/Makefile (hunk 3 above)
Comment (jochil) [reply testdata/Makefile]: I added the comment

File: integration/runner/libfuzzer/asan_test.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+	"code-intelligence.com/cifuzz/pkg/report"
+)
+
+func TestIntegration_ASAN(t *testing.T) {
+	if testing.Short() {
+		t.Skip()
+	}
+
+	utils.TestWithAndWithoutMinijail(t, func(t *testing.T, disableMinijail bool) {
+		test := utils.NewLibfuzzerTest("../../../testdata/build/trigger_asan", disableMinijail)
```
[the remark this answers is not in the document]
Comment (jochil) [reply integration/runner/libfuzzer/asan_test.go]: I added a helper in `integration/utils/filepath.go`

File: pkg/runner/libfuzzer/libfuzzer_runner.go [hunk 7]
```diff
@@ -347,6 +347,11 @@ func IsExpectedExitError(err error) bool {
 		fuzzer_runner.LibFuzzerOOMExitCode,
 		fuzzer_runner.LibFuzzerTimeoutExitCode,
 	}
+	// On macOS, the exit code is -1 when a sanitizer found something
+	// and terminates the process via SIGABRT.
```
Comment (fmeum) [inline pkg/runner/libfuzzer/libfuzzer_runner.go:351]: Strictly speaking, this is not true: Whether a sanitizer reports results in a halt or an abort depends on the value of the `abort_on_error` sanitizer option, which defaults to 1 on macOS and to 0 on Linux. We should probably either parse the ASAN_OPTIONS to determine the value, factoring in the OS-dependent default, or simply always accept both exit codes.

File: integration/utils/filepath.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func GetProjectRoot(t *testing.T) string {
+	_, filename, _, ok := runtime.Caller(0)
```
Comment (fmeum) [inline integration/utils/filepath.go:13]: AFAIR I found the working directory with `go test ./...` to be the project root, which would be a simpler way to get it.

File: pkg/runner/libfuzzer/libfuzzer_runner.go (hunk 7 above)
Comment (adombeck) [reply pkg/runner/libfuzzer/libfuzzer_runner.go:351]: We decided to set `abort_on_error=0` on macOS for now, so we don't need this code anymore.

Comment (jochil) [reply pkg/runner/libfuzzer/libfuzzer_runner.go]: I remove the code

File: integration/runner/libfuzzer/asan_test.go [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+	"testing"
+
+	"code-intelligence.com/cifuzz/integration/utils"
+	"code-intelligence.com/cifuzz/pkg/report"
+)
+
+func TestIntegration_ASAN(t *testing.T) {
+	if testing.Short() {
+		t.Skip()
+	}
```
Comment (fmeum) [inline integration/runner/libfuzzer/asan_test.go:13]: Have you tested how fast this executes if you put a `t.Parallel()` here? It did work well for the CMake and replayer tests.

File: integration/runner/libfuzzer/ubsan_non_recoverable_test.go [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"runtime"
+	"testing"
+
+	"code-intelligence.com/cifuzz/integration/utils"
+	"code-intelligence.com/cifuzz/pkg/report"
+)
+
+func TestIntegration_UBSANNonRecoverable(t *testing.T) {
+	// We are using msvc for windows with does not support msvc the way we use it
```
Comment (fmeum) [inline integration/runner/libfuzzer/ubsan_non_recoverable_test.go:12]:
```suggestion
	// We are using msvc on windows, which does not support UBSan yet.
```

File: integration/runner/libfuzzer/ubsan_recoverable_test.go [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+import (
+	"runtime"
+	"testing"
+
+	"code-intelligence.com/cifuzz/integration/utils"
+	"code-intelligence.com/cifuzz/pkg/report"
+)
+
+func TestIntegration_UBSANRecoverable(t *testing.T) {
+	// We are using msvc for windows with does not support msvc the way we use it
```
Comment (fmeum) [inline integration/runner/libfuzzer/ubsan_recoverable_test.go:12]:
```suggestion
	// We are using msvc on windows, which does not support UBSan yet.
```

File: integration/utils/filepath.go (hunk 8 above)
Comment (fmeum) [reply integration/utils/filepath.go:13]: Thinking about this some more, does it make sense to have the fuzzers in a top-level `testdata` directory? I would find it more logical to have them in `integration/testdata` as the integration tests for `cifuzz` are where they are going to be used.

File: integration/utils/minijail.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+	"runtime"
+	"testing"
+
+	"code-intelligence.com/cifuzz/pkg/runfiles"
+	"code-intelligence.com/cifuzz/tools/install"
+	"github.com/stretchr/testify/require"
+)
+
+func TestWithAndWithoutMinijail(t *testing.T, f func(t *testing.T, disableMinijail bool)) {
+	t.Run("WithoutMinijail", func(t *testing.T) {
```
Comment (fmeum) [inline integration/utils/minijail.go:13]: The subtests may also benefit from `t.Parallel()`.

File: testdata/Makefile [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+NOLINK_OPTION := -c
+OUT_OBJ_OPTION := -o
+DISABLE_UBSAN := 0
+
+# handle MSVC specific options
+ifeq ($(OS),Windows_NT)
+	CC = cl
+	CXX = cl
+	CFLAGS = -fsanitize=fuzzer \
+	  -fsanitize=address \
```
Comment (fmeum) [inline testdata/Makefile:35]: I would suggest using the MSVC flag style consistently, even if there is no functional difference.
```suggestion
	CFLAGS = /fsanitize=fuzzer \
	  /fsanitize=address \
```

File: testdata/Makefile [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+		/Zi
+
+	CXXFLAGS = -fsanitize=fuzzer \
+	  -fsanitize=address \
```
Comment (fmeum) [inline testdata/Makefile:39]:
```suggestion
	CXXFLAGS = /fsanitize=fuzzer \
	  /fsanitize=address \
```

**Review by fmeum — APPROVED** [review]
Thanks for working on and porting all the tests!

File: integration/runner/libfuzzer/asan_test.go (hunk 9 above)
Comment (jochil) [reply integration/runner/libfuzzer/asan_test.go:13]: Yes, I tried but it made the tests fail... I created a task (CLI-199) for this to investigate further

File: integration/runner/libfuzzer/ubsan_non_recoverable_test.go (hunk 10 above)
Comment (jochil) [reply integration/runner/libfuzzer/ubsan_non_recoverable_test.go:12]: done

File: integration/runner/libfuzzer/ubsan_recoverable_test.go (hunk 11 above)
Comment (jochil) [reply integration/runner/libfuzzer/ubsan_recoverable_test.go:12]: done

File: integration/utils/minijail.go (hunk 12 above)
Comment (jochil) [reply integration/utils/minijail.go:13]: I added it to CLI-199

File: integration/utils/filepath.go (hunk 8 above)
Comment (jochil) [reply integration/utils/filepath.go:13]: could make sense... these kind of test data should only be used in integration tests. In the unit tests we should mock more. I created CLI-201 to do this

File: testdata/Makefile (hunk 13 above)
Comment (jochil) [reply testdata/Makefile]: done

File: testdata/Makefile (hunk 14 above)
Comment (jochil) [reply testdata/Makefile]: done

**Review by fmeum — APPROVED** [review]

**Pull request merged**